ehrapy/api/preprocessing/_data_imputation.py

```diff
@@ -3,51 +3,300 @@
+
+    Returns:
+            The imputed AnnData object
+
+    """
+    replacement = kwargs.get("replacement")
+    imputer = SimpleImputer(strategy="most_frequent")
+    # impute a subset of columns
+    if isinstance(replacement, list):
+        column_indices = get_column_indices(adata, replacement)
+        adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+    # impute all columns if None passed
+    else:
+        adata.X = imputer.fit_transform(adata.X)
+
+    return adata
+
+
+# ===================== KNN Imputation =========================
+
+
+def _knn(adata: AnnData, **kwargs) -> AnnData:
+    """Impute data using the KNN-Imputer.
+    When using KNN Imputation with mixed data (non-numerical and numerical), encoding using ordinal encoding is required
+    since KNN Imputation can only work on numerical data. The encoding itself is just a utility and will be redone once
+    imputation ran successfully.
+
+    Args:
+        adata: The AnnData object to use KNN Imputation on
+        **kwargs: Keyword args
+                  replacement: A list of strings indicating which columns to use KNN imputation on (if None -> all columns)
+
+    Returns:
+            The imputed (but unencoded) AnnData object
+    """
+    # numerical only data needs no encoding since KNN Imputation can be applied directly
+    if np.issubdtype(adata.X.dtype, np.number):
+        _knn_impute(adata, **kwargs)
+    else:
+        # ordinal encoding is used since non-numerical data can not be imputed using KNN Imputation
+        enc = OrdinalEncoder()
+        adata.X = enc.fit_transform(adata.X)
+        # impute the data using KNN imputation
+        _knn_impute(adata, **kwargs)
+        # decode ordinal encoding to obtain imputed original data
+        adata.X = enc.inverse_transform(adata.X)
+
+    return adata
+
+
+def _knn_impute(adata: AnnData, **kwargs) -> None:
+    """Utility function to impute data using KNN-Imputer"""
+    n_neighbors = kwargs.get("n_neighbors")
+    replacement = kwargs.get("replacement")
+    imputer = KNNImputer(n_neighbors=n_neighbors if n_neighbors else 1)
+
+    if isinstance(replacement, list):
+        column_indices = get_column_indices(adata, replacement)
+        adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+    # impute all columns if None passed
+    else:
+        adata.X = imputer.fit_transform(adata.X)
 
-    # scenario 2: Replace all missing values in a subset of columns with the specified value
-    # TODO
 
-    # scenario 3: Replace all missing values in a subset of columns with a specified value per column
-    # TODO
+def _miss_forest(adata: AnnData, **kwargs) -> AnnData:
+    """Impute data using the MissForest stratgey. See https://academic.oup.com/bioinformatics/article/28/1/112/219101.
+    This requires the computation of which columns in X contain numerical only (including NaNs) and which mixed data, which could be an
```

Comment:
Do you think that we should support mixed data in the first place? Let's discuss.